neqo-http3/src/connection_client.rs

```diff
@@ -873,13 +873,13 @@ impl Http3Client {
     /// packets need to be sent or if a timer needs to be updated.
     ///
     /// [1]: ../neqo_transport/enum.ConnectionEvent.html
-    pub fn process_input(&mut self, dgram: &Datagram, now: Instant) {
-        self.process_multiple_input(iter::once(dgram), now);
+    pub fn process_input<'a>(&mut self, dgram: impl Into<Datagram<&'a [u8]>>, now: Instant) {
```

Comment:
Should this be `AsRef` rather than `Into`?  Or maybe `Borrow`.  We're not looking for a conversion.

Reply:
Unfortunately `Datagram<Vec<u8>>` can not implement `Borrow<Datagram<&[u8]>>` nor `AsRef<Datagram<&[u8]>`. 

See https://github.com/mozilla/neqo/pull/2093#discussion_r1788539953 for details and playground example.

Same for the opposite direction, i.e. `ToOwned`:

https://github.com/mozilla/neqo/pull/2093#discussion_r1788548422

Reply:
```rust
impl Borrow<[u8]> for Datagram<Vec<u8>> {
    fn borrow(&self) -> &[u8] {
        &self.d
    }
}
impl Borrow<[u8]> for Datagram<&'_ [u8]> {
    fn borrow(&self) -> &[u8] {
        self.d
    }
}

#[allow(warnings)]
fn check(_d: Datagram<impl Borrow<[u8]>>) {}
```
Seems to compile just fine.

Reply:
Or better...

```rust
impl<D: AsRef<[u8]>> Borrow<[u8]> for Datagram<D> {
    fn borrow(&self) -> &[u8] {
        self.d.as_ref()
    }
}
impl<D: AsMut<[u8]> + AsRef<[u8]>> BorrowMut<[u8]> for Datagram<D> {
    fn borrow_mut(&mut self) -> &mut [u8] {
        self.d.as_mut()
    }
}

impl Deref for Datagram
where
    Self: Borrow<[u8]>,
{
    type Target = [u8];
    #[must_use]
    fn deref(&self) -> &Self::Target {
        Borrow::<[u8]>::borrow(self)
    }
}

#[allow(warnings)]
fn test(
    _d: Datagram<impl Borrow<[u8]>>,
    _i: impl IntoIterator<Item = Datagram<impl Borrow<[u8]>>>,
) {
}
```

Reply:
Though both violate the `Borrow` contract in that you aren't getting the whole thing, just a piece of it.  `AsRef` might be a better medium for this one...

```rust
impl<D: AsRef<[u8]>> AsRef<[u8]> for Datagram<D> {
    fn as_ref(&self) -> &[u8] {
        self.d.as_ref()
    }
}
impl<D: AsMut<[u8]> + AsRef<[u8]>> AsMut<[u8]> for Datagram<D> {
    fn as_mut(&mut self) -> &mut [u8] {
        self.d.as_mut()
    }
}

impl Deref for Datagram
where
    Self: AsRef<[u8]>,
{
    type Target = [u8];
    #[must_use]
    fn deref(&self) -> &Self::Target {
        AsRef::<[u8]>::as_ref(self)
    }
}

#[allow(warnings)]
fn test(_d: Datagram<impl AsRef<[u8]>>, _i: impl IntoIterator<Item = Datagram<impl AsRef<[u8]>>>) {}
```

Reply:
Thank you. Implemented with recent commits. See https://github.com/mozilla/neqo/pull/2187#issuecomment-2426040259.